### PR TITLE
[MiniCPM-o 4.5] Split TTS into Talker and Token2Wav stages

### PR DESCRIPTION
**Pull request by CAICAIIs**
## Summary

Relates to #5184.

This PR splits the MiniCPM-o 4.5 serving TTS path from the previous fused
`llm -> tts` flow into:

```text
thinker / llm -> talker -> token2wav
```

The external OpenAI-compatible `/v1/chat/completions` behavior is preserved:
text output still comes from the thinker stage, and audio requests still return
a complete 24 kHz WAV waveform.

## Details

- Updates the MiniCPM-o 4.5 pipeline and deploy configs so Stage 1 is `talker`
  and Stage 2 is `token2wav`.
- Keeps `model_stage="tts"` as a compatibility alias for the Talker path.
- Refactors the MiniCPM-o 4.5 TTS wrapper around four explicit steps:
  prepare TTS inputs, generate audio tokens, decode audio tokens, and package
  the waveform.
- Adds Talker -> Token2Wav stage processors for scheduler sizing and full
  `codes.audio` payload handoff.
- Updates connector allowlists so the Token2Wav stage receives full payload
  inputs instead of staying in `WAITING_FOR_INPUT`.
- Adds MiniCPM-o 4.5 E2E TTS and isolated Token2Wav benchmark scripts with
  JSON/CSV output and failure-row recording.

## Scope

This PR is limited to the structural stage split and reproducible benchmark
entry points. It does not enable async chunking, change default Token2Wav
precision, add runtime tracing, tune chunk thresholds, or claim a performance
improvement.

Stage 1 still uses `max_num_seqs: 1` because the current Talker path consumes
`runtime_additional_information[0]`. Safe Talker batching should be handled in a
separate follow-up.

Missing or empty codec payloads are converted to finished empty payloads so the
downstream stage does not wait forever. Real Token2Wav init/decode failures
still fail loudly.

## Test 

GPU validation was run on a 2x RTX 5090 validation host before the final cleanup
pass:

- Text+audio E2E short benchmark at concurrency 8: 8/8 successful requests.
- Long text+audio smoke: complete 24 kHz finite waveform.
- Token2Wav benchmark: 1024 / 2048 / 4096 tokens passed in one-shot and
  streaming modes.
- 8192-token Token2Wav one-shot OOM was recorded as a failure row; 8192-token
  streaming passed.
- No `WAITING_FOR_INPUT`, serving CUDA OOM, NaN/Inf audio, or clipping was
  observed in the successful E2E runs.

### CODE REVIEW

**Comment by Sy0307** [comment]
Good thought and I wonder this idea can also be compatible for full duplex mode in #3907 (or after light modification)?

**Comment by CAICAIIs** [comment]
> Good thought and I wonder this idea can also be compatible for full duplex mode in #3907 (or after light modification)?

Yes, I think this split should also apply to #3907.

The useful boundary is Talker -> `codes.audio` -> Token2Wav waveform decode. The full-duplex runtime can keep its realtime/session logic and reuse this handoff instead of the old fused TTS+Token2Wav stage.

This PR keeps the implementation scoped to the default chat-completions complete-WAV path; realtime chunk/audio-delta handoff can be a light follow-up if needed.

**Comment by R2-Y** [comment]
there is already a PR here that splits TTS into Talker and Token2Wav stages: https://github.com/vllm-project/vllm-omni/pull/5228/changes

File: vllm_omni/model_executor/models/minicpmo_4_5/pipeline.py [hunk 1]
```diff
@@ -50,29 +44,29 @@
...
-            # Routing through the wrapper is required so that the runner-side
-            # ``runtime_additional_information`` payload reaches the talker
-            # (the standalone TTS class only reads ``additional_information``,
-            # so wiring stage 1 directly to it would always trigger the dummy
-            # path) and so the resulting waveform is packaged as
-            # ``OmniOutput.multimodal_outputs["model_outputs"]`` instead of
-            # being returned as a bare tuple that the AR runner would mistake
-            # for hidden states. ``hf_config_name="tts_config"`` keeps KV
-            # cache / mrope sizing scoped to the talker sub-config.
+            model_stage="talker",
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/minicpmo_4_5/pipeline.py:47]: 🔴 [blocking] This file needs reconciliation with current main before merge: main's version of this pipeline now registers the duplex runtime (`duplex_runtime_extension`, `duplex_serving_adapter`, `duplex_control_enabled=True`) and marks the 2-stage topology as the frozen contract the duplex runtime is built on — the streaming fused talker, the duplex deploy overlays (`minicpmo_4_5_duplex.yaml` with `cudagraph_mode: PIECEWISE` and `stop_token_ids` tuned for it), and the data-plane's cumulative-audio handling all assume stage 1 emits unit waveforms. This branch's version has none of those fields, so a rebase resolved toward the PR side silently drops duplex support, and resolved toward main contradicts the 3-stage split.

Suggestion: make the duplex story explicit here — either (a) this PR series adds a duplex-aware chunked talker→token2wav path (async_chunk + ref-audio identity in the connector payload + turn/epoch metadata on stage-2 outputs), or (b) duplex stays on the fused path for now, in which case `session_mode: duplex` should select the fused topology (or fail loudly) rather than run the 3-stage pipeline and produce no audio. Either way, the choice belongs in this file's docstring and the PR description.

File: vllm_omni/model_executor/stage_input_processors/minicpmo_4_5_omni.py [hunk 2]
```diff
@@ -107,30 +79,146 @@ def llm2tts(
...
+    function only sizes the consumer prompt so the generation runner can
+    schedule the request. Empty audio payloads still get one placeholder slot
+    and are disambiguated by ``meta.code_flat_numel=0`` in the connector
+    payload.
+    """
+    del _prompt, _requires_multimodal_data, streaming_context
+
+    token2wav_inputs: list[OmniTokensPrompt] = []
+    for talker_output in source_outputs:
+        if getattr(talker_output, "finished", True) is False:
```
Comment (linyueqian) [inline vllm_omni/model_executor/stage_input_processors/minicpmo_4_5_omni.py:164]: 🟡 [important] This skip is correct for the whole-segment PR1 contract, but it's also the exact line where duplex streaming dies after a rebase: the merged duplex runtime's talker emits ~1 s unit outputs with `finished=False` inside a long-lived request, so every unit gets discarded here and Token2Wav only ever sees the final flush — first-audio-byte latency becomes full-turn latency (talker AR up to 16 384 codes + SHM + full vocoder decode). Worth a comment here naming that constraint, and a pointer to where the async-chunk path will land in the #5184 series, so the next person doesn't discover it from a silent duplex regression.

File: vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_tts.py [hunk 3]
```diff
@@ -259,29 +279,52 @@ def generate_speech(
...
+            raise RuntimeError("MiniCPM-o 4.5 Token2Wav is not initialized")
 
         import torchaudio
 
-        model_path = self.vllm_config.model_config.model
-        default_ref = os.path.join(model_path, "assets", "HT_ref_audio.wav")
-        prompt_wav_path = default_ref if os.path.exists(default_ref) else None
+        if prompt_wav_path is None:
+            model_path = self._model_path()
+            default_ref = os.path.join(model_path, "assets", "HT_ref_audio.wav")
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/minicpmo_4_5/minicpmo_4_5_omni_tts.py:326]: 🟡 [important] Two related gaps in the split's ref-audio story:

1. The token2wav stage always lands on this bundled default: only `codes` + `meta` cross the stage boundary, and the stage-2 forward calls `decode_audio_tokens` with no `prompt_wav_path`. Per-request voice cloning therefore cannot reach the vocoder. On current main this is now load-bearing — duplex *requires* explicit `ref_audio` (`ref_audio_required` error) and conditions the in-process Token2Wav with it; after a rebase the requirement would be silently ignored here (wrong-voice audio, no error). The connector payload needs a ref-audio identity (path/hash) alongside `codes`, or an explicit statement that stage-2 voice cloning is out of scope for PR1.

2. Main's version of this file pairs a 16-entry ref-wav LRU with token2wav base-cache eviction (`_ref_audio_prompt_files` eviction pops the matching `_t2w_base_caches` entry — a worker memory-leak fix from the #3907 review cycle). This branch has neither structure, and in the split topology they'd live in different processes — the pairing needs a stage-2-local reimplementation keyed by whatever ref identity crosses the boundary.

File: vllm_omni/deploy/minicpmo_4_5_2gpu.yaml [hunk 4]
```diff
@@ -51,6 +67,24 @@ stages:
       seed: 42
       detokenize: false
 
+  - stage_id: 2
+    max_num_seqs: 1
+    gpu_memory_utilization: 0.35
+    enforce_eager: false
+    async_scheduling: false
+    max_model_len: 4096
```
Comment (linyueqian) [inline vllm_omni/deploy/minicpmo_4_5_2gpu.yaml:75]: 🟡 [important] `max_model_len: 4096` on stages 1 and 2 in this profile, but the talker's cap is `max_new_token = max(2048, min(16384, num_text*10))` and `talker2token2wav_token_only` sizes the stage-2 placeholder prompt to `numel(codes)` — so any response beyond roughly 400 text tokens (≳4 096 codes) overflows the stage-2 prompt in this profile. The 1-GPU and 3-GPU profiles use 16 384 consistently; this one looks like an oversight. Suggest 16 384 here too, or a talker-side clamp documented as a 2-GPU limitation.

**Review by linyueqian — COMMENTED** [review]
Reviewed with a focus on how this split interacts with the MiniCPM-o 4.5 full-duplex runtime that just merged in #3907 — because this PR's base predates that merge, and the merged runtime made the *fused, streaming* stage-1 talker its load-bearing substrate. First: the split itself is competently engineered along the established full-payload pattern — the producer/consumer/coordinator allowlist triple is complete, the 1-D `codes.audio` REPLACE accumulation is pinned by tests, memory separation is real (talker process lazy-inits only MiniCPMTTS, token2wav only the vocoder with a checkpoint-scan skip), and the fail-loud changes (`_lazy_init_token2wav` raising on missing assets, decode raising instead of returning None) are welcome.

The concern is the rebase. Four duplex behaviors are structurally broken by the 3-stage shape as designed (details in inline comments); two survive mechanically:

**Survives:** request identity (the orchestrator forwards to stage 2 under the same request id, and the duplex data-plane's `response_stage_id` is computed as `stage_count − 1`, so it follows automatically) and abort fan-out (duplex cancel aborts by request id across all stage pools, which reaches an existing stage-2 request).

**Breaks:**
1. **Streaming → whole-turn latency.** The duplex talker on main emits ~1 s unit waveforms inside a long-lived streaming request; this bridge only flushes on request-finish and `talker2token2wav_token_only` explicitly skips `finished=False` outputs, so under duplex the first audio byte would wait for complete talker generation (up to 16 384 codes) + SHM transfer + full vocoder decode. Understood that async chunking is scoped out of PR1 per #5184 — but the duplex runtime now ships on main, so the scoping needs to be explicit about duplex (see pipeline comment).
2. **Turn/epoch staleness filtering goes inert.** Stage-2 outputs carry only `model_outputs` + `sr` — no turn/epoch metadata — so the duplex data-plane's barge-in staleness rejection can't identify stale audio. Its segment-end detection also never fires: both of its triggers (the stop token appearing in final-stage token ids, and the `tts_is_last_chunk` output metadata) are absent from stage-2-shaped outputs.
3. **Per-request voice cloning can't reach the vocoder.** Only `codes` + `meta` cross the talker→token2wav boundary; the stage-2 decode falls back to the bundled default reference wav. Duplex on main *requires* explicit `ref_audio` — after a rebase it would be silently ignored at the vocoder (wrong-voice output with no error).
4. **The ref-audio LRU ↔ t2w base-cache eviction pairing (added on main to fix a worker memory leak) has no home in this topology** — the LRU lives with the talker, the vocoder caches live in the stage-2 process, and this PR's tree contains neither.

None of this blocks the split as a turn-based-serving refactor — but the rebase onto current main is not a mechanical conflict resolution, and "take the PR side" on `pipeline.py`/`minicpmo_4_5_omni_tts.py` would delete the duplex registration and the leak fix outright. Suggestion: state explicitly (in the PR description and `pipeline.py`) whether duplex remains on the fused stage-1 path until a later PR in the #5184 series, and if so, gate the 3-stage topology so `session_mode: duplex` deploys keep the fused path rather than silently producing no audio.

Smaller items inline: stage-1/2 `max_model_len: 4096` vs 16 384-code talker output in the 2-GPU profile, and the new empty-waveform behavior for no-TTS responses is a client-visible API change worth documenting. Also noting the request-0-only batching attribution is now duplicated into a second stage (safe only while `max_num_seqs: 1` everywhere).

Live verification note: I intended to run an A/B chat-completions parity check (fused vs split) but the shared H20/H200 test machines are mid-migration; happy to run it once they're back.

**Comment by lishunyang12** [comment]
Thanks for your contribution. Closed as stage spliting has been integrated.
